README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# FlexTable

FlexTable is a customizable table with headers, splitView, freezeView, autoFreeze, zoom and scrollbars. The table consist of a model, a viewmodel and builders. Fig. 1 shows some options. Flextable supports two-way scrolling. If the scroll direction deviates less than 22.5 degrees over a length of 30 in the vertical or horizontal direction the scroll stabiliser kicks in to prevent drifting over several pages with enthusiastic scrolling. The stabiliser corrects the drift in the cross scroll direction, so you see probability a small wobbling. The scrollbars can be used if they appear after the drag starts. (mobile)'

It is also possible to place the FlexTable in a customScrollView by wrapping the FlexTable in a adaptive sliver, wrapped in a sliver the table can only scroll in one direction at the same time.

[Web App Example](https://js-lab.nl/flextable) | [Source code](https://github.com/JoanSchi/flextable/tree/main/other_examples/table_examples)

<img src="https://js-lab.nl/flextable/flextable_options.png" width="800" >

Fig. 1: Option: **A.** scroll direction, **B.** Drag to initiate splitView, **C.** Freeze/unfreeze, **D.** Change freeze position

## Usage
To use this plugin, add flextable as a dependency in your pubspec.yaml file.


## Getting started
To make a simple flextable, make a model (DefaultFtModel) at some lines and place the model in the  DefaultFlexTable widget.


### Model
The DefaultFtModel (FtModel\<Cell\>) contains the data, lines, cell dimensions, split options, autofreeze etc. The Cell contains the value and some basic CellAttr: { textStyle, alignment, background, percentagBackground, rotate }.

```dart
  final model = DefaultFtModel(
      defaultWidthCell: 120.0,
      defaultHeightCell: 50.0,
    );


    model.addCell(row: 0, column: 0, cell: Cell(value: 'Hello!', attr: {
        
          CellAttr.textStyle: const TextStyle(
              fontSize: 20, color: Colors.blue)
     }, ));

    // Add lines (see Table Lines paragraph)
    model.horizontalLineList.addLineRange(...)
```


### FlexTable
The DefaultFlexTable (FlexTable\<FtModel\<Cell\>,Cell\>), needs the DefaultFtModel and DefaultTableBuilder. By default zoom, freeze and split (drag from: rigth,top corner) is enabled. It is possible to at FtController to FlexTable to obtain the FtViewModel to change for example scale, split, unlock etc ([flextable settings](https://github.com/JoanSchi/flextable/blob/main/other_examples/table_examples/lib/examples.dart/flextable_settings.dart),
[TableScaleSlider](https://github.com/JoanSchi/flextable/blob/main/other_examples/table_examples/lib/examples.dart/example_fruit.dart)).


```Dart
@override
  Widget build(BuildContext context) {
    return DefaultFlexTable(model: model,tableBuilder: DefaultTableBuilder(),);
  }
```

### FlexTable in CustomScrollView
FlexTable can be placed in a CustomScrollView with FlexTableToSliverBox. How does work: From FlexTableToSliverBox (RenderSliverSingleBoxAdapter) the scroll of the table can be determined with constraints.scrollOffset + overlap and the window size is determined by paintedChildSize. Two directional scrolling is not possible in CustomScrollView at the moment.

```Dart
 final ftController = DefaultFtController();

  @override
  Widget build(BuildContext context) {
    return CustomScrollView(slivers: <Widget>[
      // Other slivers...,
      FlexTableToSliverBox(
          ftController: ftController,
          child: DefaultFlexTable(
            model: model,
            tableBuilder: DefaultTableBuilder(),
            ftController: ftController,
          ))
    ]);
  }
```

### Costumize FlexTable
The Default is at the moment quite basic, nevertheless the following classes can be extended to change cell properties, cell build, table measurements, how the cell properties is received, etc:  
- FtModel\<C extends AbstractCell\> extends AbstractFtModel\<C\>
- AbstractTableBuilder\<T extends AbstractFtModel\<C\>,C extends AbstractCell\>
- FlexTable\<T extends AbstractFtModel\<C\>, C extends AbstractCell\>
- FtController\<T extends AbstractFtModel\<C\>, C extends AbstractCell\>
- AbstractCell

### Table Lines
The vertical and horizontal lines are added seperatedly to the table in ranges to minimize objects for large tables. The TableLinesOneDirection object contains a LineLinkedList with LineRanges which contains a LineLinkedList with LineNodeRanges for one direction.
The TableLinesOneDirection object will merge added LineRanges and the containing LineNodes. Equal ranges besides each other will merge into a bigger range. 
Lines or parts of the line can be deleted by adding EmptyLineNodes (LineNode with noLines) over the desired range. 
The TableLinesOneDirection will remove the lines where the EmptyLineNode are merged. If the LineRange is empty, the LineRange will be removed automatical.
A LineNodeRange or LineNode object can be reused for loops, because they are coppied or merged by TableLinesOneDirection object.

It is possible to change the width and color of the existing lines over a large range by using Line.change(width:.., color..). The new properties will merge with the existing lines.


```dart
TableLinesOneDirection horizontalLines = TableLinesOneDirection();
// or with initialized FtModel -> ftModel.horizontalLines

const blueLine = Line(
    width: 0.5,
    color: Color(0xFF42A5F5),
  );

  /// 0: --  --
  /// 1: --  --
  /// 2: --  --
  ///
  
  horizontalLines.addLineRange(LineRange(
      startIndex: 0,
      endIndex: 2,
      lineNodeRange: LineNodeRange(list: [
        LineNode(startIndex: 0, after: blueLine),
        LineNode(startIndex: 2, before: blueLine),
        LineNode(startIndex: 4, after: blueLine),
        LineNode(startIndex: 6, before: blueLine),
      ])));
 ```
Output:

LineRanges in TableLinesOneDirection: 
- LineNodeRange 0-2: 
  - LineNode 0-0: before: null, after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 2-2: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: null
  - LineNode 4-4: before: null, after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 6-6: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: null

```
  Change to:

  /// 0: --  --
  /// 1:   --
  /// 2: --  --
  ///

  horizontalLines.addLineRange(LineRange(
      startIndex: 1,
      lineNodeRange: LineNodeRange(
        list: [
          // remove line
          EmptyLineNode(startIndex: 0, endIndex: 6),
          //add new line
          LineNode(startIndex: 2, after: blueLine),
          LineNode(startIndex: 4, before: blueLine),
        ],
      )));
```
Output:

LineRanges in TableLinesOneDirection: 
- LineNodeRange 0-0: 
  - LineNode 0-0: before: null, after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 2-2: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: null
  - LineNode 4-4: before: null, after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 6-6: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: null
- LineNodeRange 1-1: 
  - LineNode 2-2: before: Line(o:LineOptions.no, w:null, c:null), after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 4-4: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: Line(o:LineOptions.no, w:null, c:null)
- LineNodeRange 2-2: 
  - LineNode 0-0: before: null, after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 2-2: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: null
  - LineNode 4-4: before: null, after: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5))
  - LineNode 6-6: before: Line(o:LineOptions.line, w:0.5, c:Color(0xff42a5f5)), after: null

```
  Change to:

  /// 0: ------
  /// 1:   --
  /// 2: ------
  ///
  /// Change all lines to: width = 2.0, color = green 

   horizontalLines.addLineRanges((add) {
    // replace complete line
    add(LineRange(
        startIndex: 0,
        lineNodeRange: LineNodeRange(
          list: [
            LineNode(
              startIndex: 0,
              after: blueLine,
            ),
            LineNode(
              startIndex: 1,
              endIndex: 5,
              before: blueLine,
              after: blueLine,
            ),
            LineNode(startIndex: 6, before: blueLine),
          ],
        )));

    // remove middle part of the line
    add(LineRange(
        startIndex: 2,
        lineNodeRange: LineNodeRange(
          list: [
            EmptyLineNode(
              startIndex: 2,
              endIndex: 4,
            ),
          ],
        )));

    // Merge changes to existing lines:
    add(LineRange(
        startIndex: 0,
        endIndex: 2,
        lineNodeRange: LineNodeRange(
          list: [
            LineNode(
                startIndex: 0,
                endIndex: 6,
                before: const Line.change(
                    width: 2.0, color: Color.fromARGB(255, 142, 212, 63)),
                after: const Line.change(
                    width: 2.0, color: Color.fromARGB(255, 142, 212, 63))),
          ],
        )));
  });
```
Output:

LineRanges in TableLinesOneDirection: 
- LineNodeRange 0-0: 
  - LineNode 0-0: before: Line(o:null, w:2.0, c:Color(0xff8ed43f)), after: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f))
  - LineNode 1-5: before: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f)), after: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f))
  - LineNode 6-6: before: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f)), after: Line(o:null, w:2.0, c:Color(0xff8ed43f))
- LineNodeRange 1-1: 
  - LineNode 2-2: before: Line(o:LineOptions.no, w:null, c:null), after: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f))
  - LineNode 4-4: before: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f)), after: Line(o:LineOptions.no, w:null, c:null)
- LineNodeRange 2-2: 
  - LineNode 0-0: before: Line(o:null, w:2.0, c:Color(0xff8ed43f)), after: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f))
  - LineNode 6-6: before: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f)), after: Line(o:null, w:2.0, c:Color(0xff8ed43f))

```
  Change to:

  /// 0: remove
  /// 1:   --
  /// 2: remove
  ///
  
  horizontalLines.addLineRanges((add) {
    // Reuse of emptyLineNodeRange to remove row 0 and 2

    final emptyLineNodeRange = LineNodeRange(
      list: [EmptyLineNode(startIndex: 0, endIndex: 6)],
    );

    for (int i in [0, 2]) {
      add(LineRange(startIndex: i, lineNodeRange: emptyLineNodeRange));
    }
  });
```

Output:
LineRanges in TableLinesOneDirection: 
- LineNodeRange 1-1: 
  - LineNode 2-2: before: Line(o:LineOptions.no, w:null, c:null), after: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f))
  - LineNode 4-4: before: Line(o:LineOptions.line, w:2.0, c:Color(0xff8ed43f)), after: Line(o:LineOptions.no, w:null, c:null)


### ToDo
- Cell press listener.
- Keep cell alive.
- Scroll to.
- More cell options and add get Widget for special cases.




CHANGELOG.md
## 0.0.2

* README.md: Fix link image, added link to other examples, removed example (already shown in tab example)

## 0.0.1

* First publish


example/lib/data/fruit.dart
import 'package:flextable/FlexTable/DataFlexTable.dart';
import 'package:flextable/FlexTable/TableItems/Cells.dart';
import 'package:flextable/FlexTable/TableLine.dart';
import 'package:flextable/FlexTable/TableModel.dart';
import 'package:flutter/material.dart';

class Fruit {
  final dataTable = DataFlexTable();
  int endTableRows = 0;
  int endTableColumns = 0;
  final fruit = [
    [
      'Soort fruit',
      2015,
      2016,
      2017,
      2018,
      2019,
    ],
    [
      'Fruit open grond (totaal)',
      19770,
      20367,
      20463,
      20443,
      20382,
    ],
    [
      'Kleinfruit open grond (totaal)',
      1613,
      1627,
      1743,
      1868,
      1859,
    ],
    [
      'Blauwe bessen',
      737,
      777,
      832,
      934,
      949,
    ],
    [
      'Bramen open grond',
      36,
      28,
      38,
      43,
      42,
    ],
    [
      'Frambozen open grond',
      151,
      202,
      250,
      259,
      252,
    ],
    [
      'Rode bessen',
      258,
      263,
      283,
      311,
      324,
    ],
    [
      'Zwarte bessen',
      352,
      277,
      234,
      224,
      206,
    ],
    [
      'Overige kleinfruit',
      79,
      80,
      106,
      98,
      85,
    ],
    [
      'Pit- en steen-vruchten (totaal)',
      17947,
      18523,
      18496,
      18334,
      18295,
    ],
    [
      'Appelrassen (totaal)',
      7600,
      7342,
      6953,
      6598,
      6421,
    ],
    [
      'Elstar',
      '',
      '',
      '',
      2892,
      2852,
    ],
    [
      'Golden Delicious',
      '',
      '',
      '',
      212,
      204,
    ],
    [
      'Jonagold (incl. Jonagored)',
      '',
      '',
      '',
      1318,
      1203,
    ],
    [
      'Junami',
      '',
      '',
      '',
      340,
      333,
    ],
    [
      'Kanzi',
      '',
      '',
      '',
      435,
      402,
    ],
    [
      'Rode Boskoop (Goudreinette)',
      '',
      '',
      '',
      333,
      311,
    ],
    [
      'Overige appelrassen',
      '',
      '',
      '',
      1068,
      1117,
    ],
    [
      'Perenrassen (totaal)',
      9234,
      9451,
      9741,
      9970,
      10086,
    ],
    [
      'Beurré Alexandre Lucas',
      '',
      '',
      '',
      673,
      646,
    ],
    [
      'Conference',
      '',
      '',
      '',
      7513,
      7570,
    ],
    [
      'Doyenné du Comice',
      '',
      '',
      '',
      763,
      771,
    ],
    [
      'Overige perenrassen',
      '',
      '',
      '',
      1022,
      1098,
    ],
    [
      'Pruimen',
      259,
      253,
      259,
      261,
      275,
    ],
    [
      'Zoete kersen',
      508,
      534,
      544,
      546,
      529,
    ],
    [
      'Zure kersen',
      327,
      290,
      270,
      245,
      251,
    ],
    [
      'Overige pit- en steenvruchten',
      527,
      1187,
      728,
      715,
      733,
    ],
    [
      'Noten',
      61,
      61,
      67,
      70,
      69,
    ],
    [
      'Wijndruiven',
      148,
      155,
      157,
      171,
      160,
    ],
    [
      'Kleinfruit onder glas (totaal)',
      64,
      87,
      95,
      123,
      101,
    ],
    [
      'Bramen onder glas',
      '',
      '',
      29,
      31,
      32,
    ],
    [
      'Frambozen onder glas',
      '',
      '',
      32,
      26,
      30,
    ],
    [
      'Overige fruit onder glas',
      '',
      '',
      33,
      66,
      39,
    ],
  ];

  Fruit() {
    int row = 2;
    int column = 0;

    dataTable.addCell(
        row: 0,
        column: 0,
        columns: 5,
        cell: Cell(
            value:
                'Fruitteelt open grond en onder glas; \n teeltoppervlakte, soort fruit',
            attr: {
              'textStyle': TextStyle(
                  color: Colors.green[900],
                  fontSize: 16.0,
                  fontWeight: FontWeight.bold),
              'textAlign': TextAlign.center,
            }));

    dataTable.addCell(
        row: 1,
        column: 1,
        columns: 4,
        cell: Cell(value: 'Perioden (ha/J)', attr: {
          'background': Colors.lime[500],
          'textStyle': const TextStyle(
              color: Colors.white, fontSize: 16.0, fontWeight: FontWeight.bold),
          'textAlign': TextAlign.center
        }));

    for (var columns in fruit) {
      int tempColumn = column;

      for (var value in columns) {
        Map attr = {
          'background': Colors.lime[row == 2 ? 500 : (row % 2 == 0 ? 50 : 100)],
          if (tempColumn == 0) 'alignment': Alignment.centerLeft,
          if (row == 2)
            'textStyle': const TextStyle(
                color: Colors.white,
                fontSize: 16.0,
                fontWeight: FontWeight.bold),
        };
        dataTable.addCell(
            row: row, column: tempColumn, cell: Cell(value: value, attr: attr));

        tempColumn++;
      }
      row++;
    }

    endTableColumns = 5;

    if (endTableRows < row) endTableRows = row;

    final v = dataTable.verticalLineList;

    final secondverticalLine = v.createLineNodeList(
      lineNode: LineNode(
        bottom: Line(color: Colors.lime[100]!, width: 1.0),
      ),
      startLevelTwoIndex: 2,
    )
      ..addLineNode(
        lineNode: LineNode(
          bottom: Line(color: Colors.lime[500]!, width: 1.0),
          top: Line(color: Colors.lime[100]!, width: 1.0),
        ),
        startLevelTwoIndex: 3,
      )
      ..addLineNode(
          startLevelTwoIndex: row,
          lineNode: LineNode(top: Line(color: Colors.lime[500]!)));

    v.addList(
        startLevelOneIndex: 1,
        endLevelOneIndex: 4,
        lineList: secondverticalLine);

    final h = dataTable.horizontalLineList;

    h.addList(
        startLevelOneIndex: 2,
        lineList: h.createLineNodeList(
            lineNode:
                LineNode(right: Line(color: Colors.lime[100]!, width: 1.0)),
            startLevelTwoIndex: 0)
          ..addLineNode(
              startLevelTwoIndex: 5,
              lineNode:
                  LineNode(left: Line(color: Colors.lime[100]!, width: 1.0))));
  }

  TableModel makeTable(
      {TargetPlatform? platform, scrollLockX = true, scrollLockY = true}) {
    double minTableScale = 0.5;
    double maxTableScale = 3.0;
    double tableScale = 1.0;

    if (platform != null) {
      switch (platform) {
        case TargetPlatform.iOS:
        case TargetPlatform.android:
        case TargetPlatform.fuchsia:
          break;
        case TargetPlatform.macOS:
        case TargetPlatform.linux:
        case TargetPlatform.windows:
          minTableScale = 1.0;
          tableScale = 1.5;
          maxTableScale = 4.0;
          break;
      }
    }

    return TableModel(
        dataTable: dataTable,
        defaultWidthCell: 80.0,
        defaultHeightCell: 50.0,
        leftPanelMargin: 4.0,
        rightPanelMargin: 4.0,
        topPanelMargin: 4.0,
        bottomPanelMargin: 4.0,
        scale: tableScale,
        minTableScale: minTableScale,
        maxTableScale: maxTableScale,
        maximumRows: endTableRows,
        maximumColumns: endTableColumns,
        scrollLockX: scrollLockX,
        scrollLockY: scrollLockY,
        specificWidth: [
          PropertiesRange(min: 0, max: 0, length: 120.0)
        ],
        specificHeight: [
          PropertiesRange(min: 0, length: 60.0),
          PropertiesRange(min: 1, max: 2, length: 40.0)
        ]);
  }
}
